code/modules/projectiles/pins.dm

```diff
@@ -0,0 +1,189 @@
+/obj/item/device/firing_pin
+	name = "electronic firing pin"
+	desc = "A small authentication device, to be inserted into a firearm receiver to allow operation. NT safety regulations require all new designs to incorporate one."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "firing_pin"
+	item_state = "pen"
+	origin_tech = "materials=2;combat=4"
+	flags = CONDUCT
+	w_class = 1
+	attack_verb = list("poked")
+	var/emagged = FALSE
+	var/fail_message = "<span class='warning'>INVALID USER.</span>"
+	var/selfdestruct = 0 // Explode when user check is failed.
+	var/force_replace = 0 // Can forcefully replace other pins.
+	var/pin_removeable = 0 // Can be replaced by any pin.
+	var/obj/item/weapon/gun/gun
+
+
+/obj/item/device/firing_pin/New(newloc)
+	..()
+	if(istype(newloc, /obj/item/weapon/gun))
+		gun = newloc
+
+/obj/item/device/firing_pin/afterattack(atom/target, mob/user, proximity_flag)
+	if(proximity_flag)
+		if(istype(target, /obj/item/weapon/gun))
+			var/obj/item/weapon/gun/G = target
+			if(G.pin && (force_replace || G.pin.pin_removeable))
+				G.pin.loc = get_turf(G)
+				G.pin.gun_remove(user)
+				to_chat(user, "<span class ='notice'>You remove [G]'s old pin.</span>")
+
+			if(!G.pin)
+				gun_insert(user, G)
+				to_chat(user, "<span class ='notice'>You insert [src] into [G].</span>")
+			else
+				to_chat(user, "<span class ='notice'>This firearm already has a firing pin installed.</span>")
+
+/obj/item/device/firing_pin/emag_act(mob/user)
+	if(!emagged)
+		emagged = TRUE
+		to_chat(user, "<span class='notice'>You override the authentication mechanism.</span>")
+
+/obj/item/device/firing_pin/proc/gun_insert(mob/living/user, obj/item/weapon/gun/G)
+	gun = G
+	forceMove(gun)
+	gun.pin = src
+	return
+
+/obj/item/device/firing_pin/proc/gun_remove(mob/living/user)
+	gun.pin = null
+	gun = null
+	return
+
+/obj/item/device/firing_pin/proc/pin_auth(mob/living/user)
+	return 1
+
+/obj/item/device/firing_pin/proc/auth_fail(mob/living/user)
+	user.show_message(fail_message, 1)
+	if(selfdestruct)
+		user.show_message("<span class='danger'>SELF-DESTRUCTING...</span><br>", 1)
+		to_chat(user, "<span class='userdanger'>[gun] explodes!</span>")
+		explosion(get_turf(gun), -1, 0, 2, 3)
+		if(gun)
+			qdel(gun)
+
+
+
+/obj/item/device/firing_pin/magic
+	name = "magic crystal shard"
+	desc = "A small enchanted shard which allows magical weapons to fire."
+
+
+// Test pin, works only near firing range.
+/obj/item/device/firing_pin/test_range
+	name = "test-range firing pin"
+	desc = "This safety firing pin allows weapons to be fired within proximity to a firing range."
+	fail_message = "<span class='warning'>TEST RANGE CHECK FAILED.</span>"
+	pin_removeable = 1
+	origin_tech = "combat=2;materials=2"
+
+/obj/item/device/firing_pin/test_range/pin_auth(mob/living/user)
+	for(var/obj/machinery/magnetic_controller/M in range(user, 3))
```

Comment:
Did you even map any of these into the test range? Maybe this needs to test for an area flag instead.

Reply:
They're mapped in, though probably still broken. I have yet to get to this yet.

Reply:
Probably the best way would be to define a new area flag: `FIRING_RANGE`.
That way you can just do something like:
```dm
var/area/A = get_area(src0
if (A && A.flags & FIRING_RANGE)
    do_things()
```

Reply:
Tbh this is bad so bad please no bad ew.

I like the idea of being able to control testing areas with magnetic controllers but either make the placement of said controller modify an area's flag, like Lohikar suggested, or maintain a global list of magnetic controllers and check that. Doing `in range()` for every time a gun is fired is lame.